src/games/strategy/triplea/delegate/MoveValidator.java

```diff
@@ -1430,8 +1430,10 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
         continue;
       }
       failureMessage = canPassThroughCanal(canalAttachment, units, player, data);
-      if (!failureMessage.isPresent()) {
```

Comment:
A bit of a shifting argument, your concern was that it was not correct.

The stream/filter version has a number of advantages:
- reduces the method in half, lot less code
- you only need to understand the filter expressions, rather than understanding the full method almost all at once.
- The filter methods are direct in their intent and what they are doing. It took some solid minutes to figure out that all we were doing was filtering values.

As evidence of it being simpler, in the filter form I would be able to tell you if the conditionals were correct or not almost immediately. But in the present form it takes far longer to determine if the conditionals are correct, let alone that we were doing a filtering operation.

Plus we get to get rid of `Optional` from the guts of the method, no more having to carry around the holder variable `failureMessage`.  We also get rid of the multiple return points.